router/internal/batch/batch.go

```diff
@@ -0,0 +1,185 @@
+			handlerOpts.HandlerSent.ServeHTTP(w, r)
+			return
+		}
+
+		bodyBytes, err := io.ReadAll(bufReader)
+		if err != nil {
+			http.Error(w, "failed to read request body", http.StatusBadRequest)
+			return
+		}
+
+		var batchOperations []json.RawMessage
+		if err = json.Unmarshal(bodyBytes, &batchOperations); err != nil {
+			// If there is an error, it's likely a malformed json array
+			// as the start character is "["
+			http.Error(w, "unexpected request body", http.StatusBadRequest)
+			return
+		}
+
+		batchOperationsLength := len(batchOperations)
+
+		addTracing(r, bodyBytes, handlerOpts.ClientHeader, batchOperationsLength, handlerOpts.BaseOtelAttributes, handlerOpts.RouterConfigVersion)
+
+		// When a max batch limit has been specified
+		if batchOperationsLength > handlerOpts.MaxEntriesPerBatch {
+			http.Error(w, "unable to process request", http.StatusBadRequest)
+			return
+		}
+
+		// Store the batch in the request context.
+		r = r.WithContext(context.WithValue(r.Context(), IsBatchedRequestKey{}, true))
+
+		// We have a batched request.
+		responses := make([]json.RawMessage, batchOperationsLength)
+
+		sem := make(chan struct{}, handlerOpts.MaxRoutines)
+		// Process each operation in parallel.
+		for i, singleOp := range batchOperations {
+			// print the pointer value of the above
+			sem <- struct{}{}
+
+			go func() {
+				defer func() {
+					<-sem
+				}()
+
+				spanCtx, span := handlerOpts.Tracer.Start(
+					r.Context(), fmt.Sprintf("batch-operation-%d", i),
+				)
+
+				defer span.End()
+
+				// TODO: Check if we should pass the spanCtx, or r.Context() to the new request
+				// Create a new request for the single operation.
+				rCopy := r.Clone(spanCtx)
+				// Reset the route context to avoid sharing mutable state.
+				rCopy = rCopy.WithContext(context.WithValue(rCopy.Context(), chi.RouteCtxKey, chi.NewRouteContext()))
+				rCopy.Body = io.NopCloser(bytes.NewBuffer(singleOp))
+
+				// Create a ResponseWriter that captures the output.
+				rw := newBufferingResponseWriter()
+				// Execute the mux handler for this operation.
+				handlerOpts.HandlerSent.ServeHTTP(rw, rCopy)
+
+				// Store the response (assuming the response is valid JSON).
+				responses[i] = rw.Body.Bytes()
+			}()
+		}
+
+		// Wait for all operations to be completed by blocking.
+		for n := handlerOpts.MaxRoutines; n > 0; n-- {
+			sem <- struct{}{}
+		}
+
+		// Drain and close the semaphore.
+		for len(sem) > 0 {
+			<-sem
+		}
+		close(sem)
+
+		// Write out the batched response as a JSON array.
+		w.Header().Set("Content-Type", "application/json")
+		if err := json.NewEncoder(w).Encode(responses); err != nil {
+			http.Error(w, "failed to encode batched response", http.StatusInternalServerError)
+		}
+	}
+
+	return http.HandlerFunc(fn)
+}
+
+func addTracing(r *http.Request, bodyBytes []byte, clientHeader config.ClientHeader, batchOperationsLength int, baseOtelAttributes []attribute.KeyValue, version string) {
```

Comment:
Can we pull this out from this package to manage this logic centralized? This package shouldn't contain telemetry logic.

Reply:
Can you elaborate a bit more, do you mean move it to tracing?

Reply:
Yes, at a place that is tracing specific. I don't see an overlap here with batching.